osu.Game/Skinning/SkinnableSound.cs

```diff
@@ -115,6 +115,10 @@ public bool Looping
         /// </summary>
         public virtual void Play()
         {
+            if (Scheduler.HasPendingTasks)
```

Comment:
This can't exist. Have you tried scheduling the Play internally instead? That would be much easier to swallow.

Reply:
If we do need to do this, a more correct path would look like this:

```diff
diff --git a/osu.Game/Skinning/SkinReloadableDrawable.cs b/osu.Game/Skinning/SkinReloadableDrawable.cs
index cef1db4bc0..1e3cdc876d 100644
--- a/osu.Game/Skinning/SkinReloadableDrawable.cs
+++ b/osu.Game/Skinning/SkinReloadableDrawable.cs
@@ -5,6 +5,7 @@
 using osu.Framework.Allocation;
 using osu.Framework.Extensions.ObjectExtensions;
 using osu.Framework.Graphics.Pooling;
+using osu.Framework.Threading;
 
 namespace osu.Game.Skinning
 {
@@ -14,6 +15,8 @@ namespace osu.Game.Skinning
     /// </summary>
     public abstract partial class SkinReloadableDrawable : PoolableDrawable
     {
+        private ScheduledDelegate? pendingSkinChange;
+
         /// <summary>
         /// Invoked when <see cref="CurrentSkin"/> has changed.
         /// </summary>
@@ -24,6 +27,15 @@ public abstract partial class SkinReloadableDrawable : PoolableDrawable
         /// </summary>
         protected ISkinSource CurrentSkin { get; private set; } = null!;
 
+        protected void FlushPendingChanges()
+        {
+            if (pendingSkinChange == null)
+                return;
+
+            pendingSkinChange.RunTask();
+            pendingSkinChange = null;
+        }
+
         [BackgroundDependencyLoader]
         private void load(ISkinSource source)
         {
@@ -31,10 +43,13 @@ private void load(ISkinSource source)
             CurrentSkin.SourceChanged += onChange;
         }
 
-        private void onChange() =>
+        private void onChange()
+        {
             // schedule required to avoid calls after disposed.
             // note that this has the side-effect of components only performing a skin change when they are alive.
-            Scheduler.AddOnce(skinChanged);
+            pendingSkinChange?.Cancel();
+            pendingSkinChange = Scheduler.Add(skinChanged);
+        }
 
         protected override void LoadAsyncComplete()
         {
diff --git a/osu.Game/Skinning/SkinnableSound.cs b/osu.Game/Skinning/SkinnableSound.cs
index 475b79053a..d7a3dcc269 100644
--- a/osu.Game/Skinning/SkinnableSound.cs
+++ b/osu.Game/Skinning/SkinnableSound.cs
@@ -115,6 +115,8 @@ public bool Looping
         /// </summary>
         public virtual void Play()
         {
+            FlushPendingChanges();
+
             samplesContainer.ForEach(c =>
             {
                 if (PlayWhenZeroVolume || c.AggregateVolume.Value > 0)

```

Note that I haven't tested this.

Reply:
Yes, I've wanted to do something similar, this however doesn't work because it updates only the `SkinnableSound` itself, and not the any of the `PoolableSkinnableSample`s. I suppose it's fine to update them in a similar fashion?

Reply:
Another thing would be to call this directly from `Player` when `RulesetSkinProvidingContainer.SourceChanged` event happens.

Reply:
> Yes, I've wanted to do something similar, this however doesn't work because it updates only the `SkinnableSound` itself, and not the any of the `PoolableSkinnableSample`s. I suppose it's fine to update them in a similar fashion?

I'm not sure I understand. You're going to have to explain yourself better.

Reply:
What I meant to say is that both `SkinnableSound` and `PoolableSkinnableSample`, which is in the `SkinnableSound` in the form of `AudioContainer` as an internal child, are both `SkinReloadableDrawable`s. Both of them therefore need to be updated before any of the samples should start playing.